improver/spotdata/height_adjustment.py

```diff
@@ -0,0 +1,217 @@
+        # be calculated for the spot index
+        coord = list(product(spot_index, thresholds))
+        needed_pair = []
+        for index, threshold in zip(spot_index, desired_thresholds):
+            needed_pair.extend(list(product([index], threshold)))
+
+        # interpolate across the cube and request needed thresholds
+        interp = LinearNDInterpolator(coord, spot_cube.data.flatten())
+        spot_data = np.reshape(interp(needed_pair), shape)
+
+        # Points outside the range of the original data return NAN. These points are replaced
+        # with the highest or lowest along the axis depending on the whether the vertical
+        # displacement was positive or negative
+        indices = np.where(np.isnan(spot_data))
+        spot_data[indices] = np.where(
+            broadcast_vertical_displacement[indices] > 0,
+            broadcast_max[indices],
+            broadcast_min[indices],
+        )
+        spot_cube.data = spot_data
+        return spot_cube
+
+    def process(self, spot_cube: Cube, neighbour: Cube,) -> Cube:
+        """
+        Adjusts spot forecast data to be relative to site height rather than
+        grid square orography height.
+
+        Args:
+            spot_cube:
+                A cube of spot forecasts. If this is a cube of probabilities
+                then the units of the threshold coordinate must be convertible to
+                metres as this is expected to represent a vertical coordinate. There
+                must be at least two thresholds on this cube.
+                If this is a cube of percentiles or realizations then the
+                units of the cube must be convertible to metres as the cube is
+                expected to represent a vertical profile.
+            neighbour:
+                A cube containing information about the spot data sites and
+                their grid point neighbours.
+        Returns:
+            A cube of the same shape as spot_data but with data adjusted to account for
+            the difference between site height and orography height.
+
+        Raises:
+            ValueError:
+                If spot_cube is a probability cube and there are fewer than two thresholds.
+        """
+        vertical_displacement = extract_vertical_displacements(
+            neighbour_cube=neighbour,
+            neighbour_selection_method_name=self.neighbour_selection_method,
+        )
+
+        if is_probability(spot_cube):
+            threshold_coord = find_threshold_coordinate(spot_cube)
+            self.threshold_coord = threshold_coord.copy()
+
+            if len(self.threshold_coord.points) < 2:
+                raise ValueError(
+                    f"There are fewer than 2 thresholds present in this cube, these are "
+                    f"{spot_cube.coord(threshold_coord).points}. At least two thresholds "
+                    "are needed for interpolation"
```

Comment:
As the criteria here is for fewer than two thresholds, the number of thresholds must actually be exactly one, to trigger this error message, so should this be:
```suggestion
                    "There are fewer than 2 thresholds present in this cube, this is "
                    f"{spot_cube.coord(threshold_coord).points}. At least two thresholds "
                    "are needed for interpolation"
```

Reply:
Done